src/groth16/prover.rs

```diff
@@ -293,150 +317,245 @@ where
-            drop(c);
-            a.divide_by_z_on_coset(&worker);
-            a.icoset_fft(&worker, &mut fft_kern)?;
-            let mut a = a.into_coeffs();
-            let a_len = a.len() - 1;
-            a.truncate(a_len);
-
-            Ok(Arc::new(
-                a.into_iter().map(|s| s.0.into_repr()).collect::<Vec<_>>(),
-            ))
-        })
-        .collect::<Result<Vec<_>, SynthesisError>>()?;
+    info!("a_s");
+    let provers_len = provers.len();
+
+    let params = &params;
+    let worker = &worker;
+    let provers_ref = &mut provers;
+    par! {
+        let a_s = {
+            provers_ref
+                .iter_mut()
+                .map(|prover| {
+                    let mut a = EvaluationDomain::from_coeffs(std::mem::replace(
+                        &mut prover.a,
+                        Vec::new(),
+                    ))?;
+                    let mut b = EvaluationDomain::from_coeffs(std::mem::replace(
+                        &mut prover.b,
+                        Vec::new(),
+                    ))?;
+                    let mut c = EvaluationDomain::from_coeffs(std::mem::replace(
+                        &mut prover.c,
+                        Vec::new(),
+                    ))?;
+
+                    info!("a: ifft");
+                    a.ifft(&worker, &mut fft_kern)?;
+                    a.coset_fft(&worker, &mut fft_kern)?;
+
+                    info!("b: ifft");
+                    b.ifft(&worker, &mut fft_kern)?;
+                    b.coset_fft(&worker, &mut fft_kern)?;
+
+                    info!("c: ifft");
+                    c.ifft(&worker, &mut fft_kern)?;
+                    c.coset_fft(&worker, &mut fft_kern)?;
+
+                    info!("fft collect");
+
+                    a.mul_assign(&worker, &b);
+                    a.sub_assign(&worker, &c);
+
+                    drop(b);
+                    drop(c);
+
+                    info!("coset");
+                    a.divide_by_z_on_coset(&worker);
+                    a.icoset_fft(&worker, &mut fft_kern)?;
+
+                    let a = a.into_coeffs();
+                    let a_len = a.len() - 1;
+
+                    Ok(Arc::new(
+                        a.into_iter()
+                         .take(a_len)
+                         .map(|s| s.0.into_repr())
+                         .collect::<Vec<_>>(),
+                    ))
+                })
+                .collect::<Result<Vec<_>, SynthesisError>>()
+        },
+        let params_h = {
+            params.get_h(n)
+        },
+        let params_l = {
+            params.get_l(provers_len)
+        },
+        let a_source = {
+            params.get_a(input_len, a_aux_density_total)
+        },
+        let b_g1_source = {
+            params.get_b_g1(b_input_density_total, b_aux_density_total)
+        },
+        let b_g2_source = {
+            params.get_b_g2(b_input_density_total, b_aux_density_total)
+        }
+    };
 
```

Comment:
The `fft_kern` used to get `drop()`ed here. Is not dropping it anymore intentional?

Reply:
This question is still open.

Reply:
it gets auto dropped because it is moved into the spawned thread that does the ffts